acvm-repo/acvm/src/compiler/optimizers/merge_expressions.rs

```diff
@@ -77,40 +77,38 @@ impl<F: AcirField> MergeExpressionsOptimizer<F> {
             for w in witnesses {
                 // We do not simplify circuit inputs and outputs
                 if !circuit_io.contains(&w) {
-                    used_witness.entry(w).or_default().insert(i);
+                    used_witnesses.entry(w).or_default().insert(i);
                 }
             }
         }
 
         // For each opcode, try to get a target opcode to merge with
-        for (i, opcode) in circuit.opcodes.iter().enumerate() {
+        for (op1, opcode) in circuit.opcodes.iter().enumerate() {
             if !matches!(opcode, Opcode::AssertZero(_)) {
                 continue;
             }
-            if let Some(opcode) = self.get_opcode(i, circuit) {
+            if let Some(opcode) = self.get_opcode(op1, circuit) {
                 let input_witnesses = self.witness_inputs(&opcode);
                 for w in input_witnesses {
-                    let Some(gates_using_w) = used_witness.get(&w) else {
+                    let Some(gates_using_w) = used_witnesses.get(&w) else {
                         continue;
                     };
                     // We only consider witness which are used in exactly two arithmetic gates
                     if gates_using_w.len() == 2 {
                         let first = *gates_using_w.first().expect("gates_using_w.len == 2");
                         let second = *gates_using_w.last().expect("gates_using_w.len == 2");
-                        let b = if second == i {
+                        let op2 = if second == op1 {
                             first
                         } else {
                             // sanity check
-                            assert!(i == first);
+                            assert!(op1 == first);
                             second
                         };
+
                         // Merge the opcode with smaller index into the other one
-                        // by updating modified_gates/deleted_gates/used_witness
+                        // by updating modified_gates/deleted_gates/used_witnesses
                         // returns false if it could not merge them
-                        let mut merge_opcodes = |op1, op2| -> bool {
-                            if op1 == op2 {
-                                return false;
-                            }
+                        if op1 != op2 {
```

Comment:
nit: we could actually lean into the `merge_opcodes` closure and instead turn the later if statements into early returns to reduce nesting.